### PR TITLE
[MCC-1809] Mock LedgerSyncService

### PR DESCRIPTION
**Pull request by mfaulk**
### Motivation

This is a stepping stone towards finishing https://github.com/mobilecoinofficial/mobilecoin/pull/449

ByzantineLedger depends on an instance of LedgerSyncService. This makes ByzantineLedger hard to unit test because LedgerSyncService is a complex thing that makes network requests. ByzantineLedger would be easier to unit test if LedgerSyncService could be mocked...

### In this PR
* Creates a mockable LedgerSync trait
* ByzantineLedgerWorker uses the LedgerSync trait, which allows it to use a mock. 
* Moves some files into subdirectories for easier navigation.

A few of the changes here related to generic arguments in trait methods were done in order to fit within the limitations of the mockall library. There's probably room to finesse things further, but I think it worked out OK.

### CODE REVIEW

File: consensus/service/src/byzantine_ledger/worker.rs [hunk 1]
```diff
@@ -88,7 +88,8 @@ pub struct ByzantineLedgerWorker<
     network_state: SCPNetworkState,
 
     // Ledger sync service
-    ledger_sync_service: LedgerSyncService<L, PC, ReqwestTransactionsFetcher>,
+    // ledger_sync_service: LedgerSyncService<L, PC, ReqwestTransactionsFetcher>,
+    ledger_sync_service: Box<dyn LedgerSync<SCPNetworkState>>,
```
Comment (mfaulk) [inline consensus/service/src/byzantine_ledger/worker.rs:92]: This is the end goal: ledger_sync_service can now accept an instance of MockLedgerSync.

File: ledger/sync/src/ledger_sync/ledger_sync_trait.rs [hunk 2]
```diff
@@ -0,0 +1,20 @@
+// Copyright (c) 2018-2020 MobileCoin Inc.
+
+use crate::{LedgerSyncError, NetworkState};
+use mockall::*;
+
+#[automock]
+pub trait LedgerSync<NS: NetworkState + 'static> {
```
Comment (mfaulk) [inline ledger/sync/src/ledger_sync/ledger_sync_trait.rs:7]: Adding this trait and mocking it is the main change.

File: ledger/sync/src/ledger_sync/ledger_sync_trait.rs [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+    /// Returns true if the local ledger is behind the network's consensus view of the ledger.
+    fn is_behind(&self, network_state: &NS) -> bool;
+
+    /// Attempts to synchronize the local ledger with the consensus view of the network.
+    /// # Arguments
+    /// * `network_state` - Current state of the network.
+    /// * `limit` - Maximum number of blocks to add to the ledger.
+    fn attempt_ledger_sync(
+        &mut self,
+        network_state: &NS,
```
Comment (mfaulk) [inline ledger/sync/src/ledger_sync/ledger_sync_trait.rs:17]: This used to be `network_state: &impl NetworkState`, but mockall is unable to deal with that.

File: ledger/sync/src/ledger_sync/ledger_sync_service.rs [hunk 4]
```diff
@@ -34,27 +34,19 @@ const DEFAULT_GET_BLOCKS_TIMEOUT: Duration = Duration::from_secs(30);
 const DEFAULT_GET_TRANSACTIONS_TIMEOUT: Duration = Duration::from_secs(30);
 
 pub struct LedgerSyncService<L: Ledger, BC: BlockchainConnection, TF: TransactionsFetcher> {
-    /// Local ledger database.
```
Comment (eranrund) [inline ledger/sync/src/ledger_sync/ledger_sync_service.rs:37]: Why did some comments get removed and one got to stay?

**Review by eranrund — APPROVED** [review]

File: ledger/sync/src/ledger_sync/ledger_sync_service.rs (hunk 4 above)
Comment (mfaulk) [reply ledger/sync/src/ledger_sync/ledger_sync_service.rs:37]: Just a caffeinated impulse. Some of the comments felt like tautologies, e.g.:

```
/// Transactions Fetcher.	
    transactions_fetcher: Arc<TF>
```

**Review by sugargoat — APPROVED** [review]

**Pull request merged**